File: src/main/java/superstitio/delayHpLose/DelayHpLosePower_ApplyOnAttacked.java
package superstitio.delayHpLose;

import com.megacrit.cardcrawl.cards.DamageInfo;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.core.AbstractCreature;
import com.megacrit.cardcrawl.powers.AbstractPower;
import superstitio.DataManager;

import java.util.Optional;

public class DelayHpLosePower_ApplyOnAttacked extends DelayHpLosePower{
    public static final String POWER_ID = DataManager.MakeTextID(DelayHpLosePower_ApplyOnAttacked.class);

    public DelayHpLosePower_ApplyOnAttacked(final AbstractCreature owner, int amount) {
        super(POWER_ID, owner, amount);
    }

    @Override
    public boolean checkShouldInvisibleTips() {
        return false;
    }

    @Override
    public void onInflictDamage(DamageInfo info, int damageAmount, AbstractCreature target) {
        if (info.type != DamageInfo.DamageType.NORMAL) return;
        Optional<AbstractPower> preventHpLimit = owner.powers.stream().filter(power -> power instanceof IPreventHpLimit).findAny();
        Optional<IPreventHpLimit> iPreventHpLimit = preventHpLimit.map(power -> (IPreventHpLimit) power);
        if (!preventHpLimit.isPresent()) {
            immediate_applyDamage(this);
            return;
        }
        else if (preventHpLimit.get().amount == 0) {
            immediate_applyDamage(this);
            return;
        }
        else {
            iPreventHpLimit.ifPresent(IPreventHpLimit::onPreventHpLimit);
        }
        return;
    }


    @Override
    public void onVictory() {
        immediate_applyDamage(this);
    }

    @Override
    protected int addToBot_removeDelayHpLoss(int amount, boolean removeOther) {
        int lastAmount = amount - this.amount;
        addToBot_reducePowerToOwner(this.ID, amount);
        return lastAmount;
    }

    @Override
    public boolean showDecreaseAmount() {
        return true;
    }

    public interface IPreventHpLimit {
        void onPreventHpLimit();
    }
}

File: superstitioMain/java/superstitio/delayHpLose/DelayHpLosePower_ApplyOnAttacked.kt
package superstitio.delayHpLose

import com.megacrit.cardcrawl.cards.DamageInfo
import com.megacrit.cardcrawl.cards.DamageInfo.DamageType
import com.megacrit.cardcrawl.core.AbstractCreature
import com.megacrit.cardcrawl.powers.AbstractPower
import superstitio.DataManager
import superstitioapi.utils.addToBot_reducePowerToOwner

class DelayHpLosePower_ApplyOnAttacked(owner: AbstractCreature, amount: Int) :
    DelayHpLosePower(POWER_ID, owner, amount)
{
    override fun checkShouldInvisibleTips(): Boolean
    {
        return false
    }

    override fun wasHPLost(info: DamageInfo, damageAmount: Int)
    {
        if (info.type != DamageType.NORMAL)
            return
        val preventHpLimit =
            owner.powers.firstOrNull { power: AbstractPower? -> power is IPreventHpLimit }
        if (preventHpLimit == null)
        {
            immediate_applyDamage(this)
            return
        }
        if (preventHpLimit.amount == 0)
        {
            immediate_applyDamage(this)
            return
        }
        preventHpLimit.let { (it as IPreventHpLimit).onPreventHpLimit() }
        return
    }


    override fun onVictory()
    {
//        immediate_applyDamage(this);
    }

    override fun showDecreaseAmount(): Boolean
    {
        return true
    }

    override fun addToBot_removeDelayHpLoss(amount: Int, removeOther: Boolean): Int
    {
        val lastAmount = amount - this.amount
        addToBot_reducePowerToOwner(this.ID, amount)
        return lastAmount
    }

    interface IPreventHpLimit
    {
        fun onPreventHpLimit()
    }

    companion object
    {
        val POWER_ID: String = DataManager.MakeTextID(DelayHpLosePower_ApplyOnAttacked::class.java)
    }
}
